### PR TITLE
Fix: handle version comparison corner cases in Ubuntu packages

### LINKED ISSUE #2531: fwupd-signed marqued as upgradable for Ubuntu 20.04 clients on Uyuni UI but not through apt
Uyuni 2020.07

I have observed in the Uyuni UI this package flagged as upgradable for an Uyuni 20.04 client:

```
 	Latest Package 	                        Installed Package
	fwupd-signed-1.27+1.3.9-4.amd64-deb 	fwupd-signed-1.27.1ubuntu2+1.3.11-1~focal1.amd64-deb 
```

But not listed with `apt update`:

```
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Calculating upgrade... Done
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
```

### PR DESCRIPTION
**Pull request by mbologna**
## What does this PR change?

Handle version comparison corner cases in Ubuntu packages

## GUI diff

No difference.

- [X] **DONE**

## Documentation
- No documentation needed: internal

- [X] **DONE**

## Test coverage
- Unit tests and database tests were added

- [X] **DONE**

## Links

Fixes #2531

- [X] **DONE**

## Changelogs

If you don't need a changelog check, please mark this checkbox:

- [ ] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_lint_checkstyle"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "schema_migration_test_oracle"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

File: java/code/src/com/redhat/rhn/common/util/test/RpmVersionComparatorTest.java [hunk 1]
```diff
@@ -168,6 +168,11 @@ public void testUbuntuBugzilla1150113() {
     public void testBugzilla1173201() {
         assertCompareSymm(-1, "1.27+1.3.9", "1.27.1ubuntu1+1.3.9");
         assertCompareSymm(-1, "1.27+1.3.9", "1.27.1+1.3.9");
+        assertCompareSymm(-1, "1.27+1.3.9", "1.27.1ubuntu2+1.3.11");
+        assertCompareSymm(1, "1.27+1.3.9", "1.3.11");
+        assertCompareSymm(1, "1.27", "1.3.11");
+        assertCompareSymm(1, "1.27+1.3.9", "1.3.9");
+        assertCompareSymm(-1, "1.27+1.3.9", "5.18.4.1");
```
Comment (mcalmer) [inline java/code/src/com/redhat/rhn/common/util/test/RpmVersionComparatorTest.java:175]: You adapted the check to test explicit of `~` not found, but no of this tests here has a case where a tilde exists.
Can you add such a test?

File: schema/susemanager-schema-4.2.0-to-susemanager-schema-4.2.1/100-rpm.pkb.sql [hunk 2]
```diff
@@ -0,0 +1,255 @@
+-- oracle equivalent source sha1 539cb03eb177b7e87992701071488bbb32bb0624
```
Comment (mcalmer) [inline schema/susemanager-schema-4.2.0-to-susemanager-schema-4.2.1/100-rpm.pkb.sql:1]: Not sure if this should go into the uyuni release which is currently prepared.

If yes, please change the number of the script as we already have a `100-...` .
If not, you need to change the directory as it will be released for 4.2.1-to-4.2.2

**Review by mcalmer — CHANGES_REQUESTED** [review]

File: schema/susemanager-schema-4.2.0-to-susemanager-schema-4.2.1/100-rpm.pkb.sql (hunk 2 above)
Comment (mbologna) [reply schema/susemanager-schema-4.2.0-to-susemanager-schema-4.2.1/100-rpm.pkb.sql:1]: You are right - this time I deducted the right dir but I was not considering the release. Adjusting with the nextver dir.

File: java/code/src/com/redhat/rhn/common/util/test/RpmVersionComparatorTest.java (hunk 1 above)
Comment (mbologna) [reply java/code/src/com/redhat/rhn/common/util/test/RpmVersionComparatorTest.java:175]: Fixed

File: schema/susemanager-schema-4.2.0-to-susemanager-schema-4.2.1/100-rpm.pkb.sql (hunk 2 above)
Comment (juliogonzalez) [reply schema/susemanager-schema-4.2.0-to-susemanager-schema-4.2.1/100-rpm.pkb.sql:1]: @mcalmer I'd say 4.2.0-to-4.2.1. The current spec on master is '4.2.0' and I still didn't tag for the Uyuni 2020.09 release.

What am I missing?

Comment (juliogonzalez) [reply schema/susemanager-schema-4.2.0-to-susemanager-schema-4.2.1/100-rpm.pkb.sql:1]: Maybe you were thinking about the CVE? We'll not do a full release for that. Instead we'll release a set of packages to be installed on top of Uyuni 2020.07: https://github.com/uyuni-project/uyuni/wiki/Releasing-Uyuni-patches

This is nothing we are going to do frequently at all, but I created the mechanism for important situations such as this CVE.

**Review by mcalmer — APPROVED** [review]
We already have a case where RPM version compare is broken. This does not make it worse.
So we can merge it but need to work on a real solution.

**Comment by mbologna** [comment]
> We already have a case where RPM version compare is broken. This does not make it worse.
> So we can merge it but need to work on a real solution.

```
assertCompareSymm(-1, "1.27+1.3.9", "1.27.1+1.3.9");
$> LANG=C zypper versioncmp 1.27+1.3.9 1.27.1+1.3.9
1.27+1.3.9 is newer than 1.27.1+1.3.9

assertCompareSymm(1, "1.27", "1.3.11");
$> LANG=C zypper versioncmp 1.27 1.3
1.27 is newer than 1.3
```

**Comment by mcalmer** [comment]
`-1` is older, `1` is newer.

The first example in debian should be older (-1) while in rpm it is newer (1)
